azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,388 @@ def _run_op(self, op):
+                )
+            )
+            logger.error("response body: {}".format(provisioning_op.response_body))
+            return exceptions.ServiceError(
+                "{prov_op_name} request returned a service error status code {status_code}".format(
+                    prov_op_name=prov_op_name, status_code=provisioning_op.status_code
+                )
+            )
+        else:
+            return None
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _decode_response(self, provisioning_op):
+        return json.loads(provisioning_op.response_body.decode("utf-8"))
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_registration_status(self, decoded_response):
+        return get_optional_element(decoded_response, "status")
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_operation_id(self, decoded_response):
+        return get_optional_element(decoded_response, "operationId")
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _form_complete_result(self, operation_id, decoded_response, status):
+        """
+        Create the registration result from the complete decoded json response for details regarding the registration process.
+        """
+        decoded_state = (get_optional_element(decoded_response, "registrationState"),)
+        registration_state = None
+        if decoded_state is not None:
+            # Everything needs to be converted to string explicitly for python 2
+            # as everything is by default a unicode character
+
+            registration_state = RegistrationState(
+                device_id=get_optional_element(decoded_state[0], "deviceId"),
+                assigned_hub=get_optional_element(decoded_state[0], "assignedHub"),
+                sub_status=get_optional_element(decoded_state[0], "substatus"),
+                created_date_time=get_optional_element(decoded_state[0], "createdDateTimeUtc"),
+                last_update_date_time=get_optional_element(
+                    decoded_state[0], "lastUpdatedDateTimeUtc"
+                ),
+                etag=get_optional_element(decoded_state[0], "etag"),
+                payload=get_optional_element(decoded_state[0], "payload"),
+            )
+
+        registration_result = RegistrationResult(
+            operation_id=operation_id, status=status, registration_state=registration_state
+        )
+        return registration_result
+
+
+class PollingStatusStage(CommonProvisioningStage):
+    """
+    This stage is responsible for sending the query request once initial response
+    is received from the registration response.
+    Upon the receipt of the response this stage decides whether
+    to send another query request or complete the procedure.
+    """
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _run_op(self, op):
+        if isinstance(op, pipeline_ops_provisioning.SendQueryRequestOperation):
+            query_status_op = op
+
+            def on_query_response(op, error):
```

Comment:
>def on_query_response(op, error): [](start = 12, length = 33)

It's kinda odd that this functionality essentially is duplicated on two different stages. I see why, but this doesn't seem an ideal solution because it effectively means any changes have to be made and tested in two different places.

Reply:
i initially attempted to have a common function in the super common stage to handle response. But the responses of the 2 stages are only similar in some respect and not for all the scenarios. Even if defined a common function I will need to override. 

---
In reply to: [365417577](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r365417577) [](ancestors = 365417577)